batch.py

```diff
@@ -0,0 +1,324 @@
+    )
+
+    # export this as a Hail table for downstream analysis
+    ht_path = output_path(f"{gene_name}_rare_promoter_summary.ht")
+    ht = mt.rows()
+    ht.write(ht_path)
+
+    # export MT object to PLINK (promoter variants)
+    plink_path = output_path(f"{plink_output_prefix}/{gene_name}")
+    export_plink(mt, plink_path, ind_id=mt.s)
+
+    return plink_path
+
+
+# endregion GET_GENE_SPECIFIC_VARIANTS
+
+# region MISCELLANEOUS
+
+
+def make_gene_loc_dict(file) -> dict[str, dict]:
+    """
+    Turn gene information into a dictionary
+    to avoid opening this file for every gene
+    """
+    gene_dict = {}
+
+    with open(file) as handle:
+        reader = DictReader(handle, delimiter="\t")
+
+        for row in reader:
+            gene_dict[row["gene_name"]] = row
+
+    return gene_dict
+
+
+# copied from https://github.com/populationgenomics/tob-wgs/blob/main/scripts/eqtl_hail_batch/launch_eqtl_spearman.py
+# check whether it needs modifying
+def remove_sc_outliers(df):
+    """
+    Remove outlier samples, as identified by sc analysis
+    """
+    outliers = ["966_967", "88_88"]
+    df = df[-df.sampleid.isin(outliers)]
+
+    return df
+
+
+# endregion MISCELLANEOUS
+
+config = get_config()
+
+
+@click.command()
+@click.option("--chromosomes")
+@click.option("--genes")
+@click.option("--celltypes")
+@click.option("--expression-file-prefix")
+@click.option("--sample-mapping-file")
+@click.option("--output-path")
+@click.option("--mt-path")
+@click.option("--anno-ht-path")
+def crm_pipeline(
+    chromosomes: list[str],
+    genes: list[str],
+    celltypes: list[str],
+    expression_files_prefix: str,  # 'scrna-seq/grch38_association_files'
+    sample_mapping_file: str,
+    output_path: str,  # 'tob_wgs_rv/pseudobulk_rv_association'
+    mt_path: str = DEFAULT_JOINT_CALL_MT,  # 'mt/v7.mt'
+    anno_ht_path: str = DEFAULT_ANNOTATION_HT,  # 'tob_wgs_vep/104/vep104.3_GRCh38.ht'
+    window_size: int = 50000,
+):
+
+    sb = hb.ServiceBackend(
+        billing_project=config["hail"]["billing_project"],
+        remote_tmpdir=remote_tmpdir(),
+    )
+    batch = hb.Batch("CellRegMap pipeline", backend=sb)
+
+    # extract samples for which we have single-cell (sc) data
+    sample_mapping_file = remove_sc_outliers(sample_mapping_file)
+    sc_samples = sample_mapping_file["InternalID"].unique()
+
+    # filter to QC-passing, rare, biallelic variants
+    filter_job = batch.new_python_job(name="MT filter job")
+    output_mt_path = output_path("tob_wgs_rv/densified_rv_only.mt")
+    filter_job.image(CELLREGMAP_IMAGE)
+    filter_job.call(
+        filter_variants,
+        mt_path=mt_path,
+        samples=sc_samples,
+        output_mt_path=output_mt_path,
+    )
+
+    # grab all relevant genes across all chromosomes
+    # simpler if gene details are condensed to one file
+    gene_dict: dict[str, dict] = {}
+    for chromosome in chromosomes:
+        geneloc_tsv_path = os.path.join(
```

Comment:
Could you fetch the gene intervals directly from GENCODE instead and just pass a list of gene IDs?

Reply:
i presume the file(s) i am using here do come from gencode already, but I am not sure what you mean exactly?

Reply:
@lgruen do you mean polling the [Ensembl REST API](https://rest.ensembl.org/documentation/info/symbol_lookup), for example?

```https://rest.ensembl.org/lookup/symbol/homo_sapiens/BRCA2?content-type=application/json```